src/python/py/models/builder.py

```diff
@@ -2248,8 +2313,89 @@ def make_mlp_fc(self, layer_id, mlp, root_input):
+            kwargs["weight2"], kwargs.get("bias2", ""),
+            kwargs.get("weight3", ""), kwargs.get("bias3", ""),
+        ]
+        output = f"{name}/output_0"
+
+        extra_kwargs = {"swiglu_limit": self.moe_attrs["swiglu_limit"]} if self.moe_attrs["swiglu_limit"] is not None else {}
+        self.make_node(
+            "MoE", inputs=inputs, outputs=[output], name=name, domain="com.microsoft",
+            activation_alpha=self.moe_attrs["activation_alpha"],
+            activation_beta=self.moe_attrs["activation_beta"],
+            activation_type=self.moe_attrs["activation_type"],
+            k=self.moe_attrs["top_k"],
+            normalize_routing_weights=self.moe_attrs["normalize_routing_weights"],
+            swiglu_fusion=self.moe_attrs["swiglu_fusion"],
+            use_sparse_mixer=self.moe_attrs["use_sparse_mixer"],
+            **extra_kwargs,
+        )
+        self.make_value(output, self.io_dtype, shape=['batch_size', 'sequence_length', self.hidden_size])
+
+    def make_qmoe_op(self, name, **kwargs):
+        inputs = [
+            kwargs["root_input"], kwargs["router_probs"],
+            kwargs["weight1"], kwargs["scales1"], kwargs.get("bias1", ""),
+            kwargs["weight2"], kwargs["scales2"], kwargs.get("bias2", ""),
+            kwargs.get("weight3", ""), kwargs.get("scales3", ""), kwargs.get("bias3", ""),
+        ]
+        output = f"{name}/output_0"
+
+        extra_kwargs = {"swiglu_limit": self.moe_attrs["swiglu_limit"]} if self.moe_attrs["swiglu_limit"] is not None else {}
+        self.make_node(
+            "QMoE", inputs=inputs, outputs=[output], name=name, domain="com.microsoft",
+            activation_alpha=self.moe_attrs["activation_alpha"],
+            activation_beta=self.moe_attrs["activation_beta"],
+            activation_type=self.moe_attrs["activation_type"],
+            expert_weight_bits=self.moe_attrs["expert_weight_bits"],
+            k=self.moe_attrs["top_k"],
+            normalize_routing_weights=self.moe_attrs["normalize_routing_weights"],
+            swiglu_fusion=self.moe_attrs["swiglu_fusion"],
+            use_sparse_mixer=self.moe_attrs["use_sparse_mixer"],
+            **extra_kwargs,
+        )
+        self.make_value(output, self.io_dtype, shape=['batch_size', 'sequence_length', self.hidden_size])
+
+    def make_qmoe_weights(self, weights):
+        dtype = torch.quint4x2 if self.moe_attrs["expert_weight_bits"] == 4 else torch.int8
+        qweight, scales = None, None
+
+        unsuccessful = True
+        try:
+            import tensorrt_llm
+
+            _, qweight, scales = (
+                torch.ops.trtllm._symmetric_quantize_last_axis_of_batched_matrix(weights.detach().cpu().contiguous(), dtype)
+            )
+            unsuccessful = False
+        except ImportError:
+            print("WARNING: TensorRT-LLM is needed to use torch.ops.trtllm._symmetric_quantize_last_axis_of_batched_matrix().")
+        except RuntimeError as r:
```

Comment:
```suggestion
        except RuntimeError as e:
```

nit: prefer e for errors?

Reply:
I used `r` for `RuntimeError` because `err` contains the actual error later on.